File: leetcode/graph/207. Course Schedule/py/main.py
import collections


# Time O(E + V) where |V| is the number of courses, and |E| is the number of dependencies.
# Space O(E + V)
class Solution:
    def canFinish(self, numCourses, prerequisites):
        def dfs(u):
            if color[u] == 2:
                return False

            color[u] = 1

            for v in adj_list[u]:
                if color[v] == 1:
                    return True

                if color[v] == 0 and dfs(v):
                    return True

            color[u] = 2

            return False

        # 0 - not used
        # 1 - processing
        # 2 - processed
        color = [0] * numCourses

        adj_list = collections.defaultdict(list)

        for u, v in prerequisites:
            adj_list[u].append(v)

        for course in range(numCourses):
            if dfs(course):
                return False

        return True

File: leetcode/graph/207. Course Schedule/js/main.js
// Time O(N + E)
// Space O(N)
const canFinish = (numCourses, prerequisites) => {
  let adjList = [];

  for (let i = 0; i < numCourses; i++) {
    adjList[i] = [];
  }

  // создаем структуру [id курса] -> [массив курсов необходимых для этого курса]
  prerequisites.forEach(([u, v]) => adjList[u].push(v));

  let visited = [];
  let stack = [];

  for (let i = 0; i < numCourses; i++) {
    if (hasCycle(i)) {
      return false;
    }
  }

  return true;

  function hasCycle(u) {
    if (visited[u]) {
      return false;
    }
    visited[u] = true;
    stack[u] = true;

    for (let v of adjList[u]) {
      if (stack[v]) {
        return true;
      }

      if (!visited[v] && hasCycle(v)) {
        return true;
      }
    }

    // Backtracking
    stack[u] = false;

    return false;
  }
};
